File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionFindIndex.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;

import java.util.Iterator;
import java.util.Map;

public class TransformerFunctionFindIndex extends TransformerFunction {
    public TransformerFunctionFindIndex() {
        super(FunctionDescription.of(
            Map.of(
            "by", ArgumentType.of(ArgType.Transformer).position(0).defaultIsNull(true)
            )));
    }
    @Override
    public Object apply(FunctionContext context) {
        var streamer = context.getJsonElementStreamer(null);
        if (streamer == null)
            return null;
        var hasBy = context.has("by");
        var by = context.getJsonElement("by", false);
        var adapter = context.getAdapter();
        var i = 0;
        boolean found = false;
        for (Iterator<?> iter = streamer.stream().iterator(); iter.hasNext();) {
            var item = iter.next();
            if (!hasBy) {
                if (adapter.isTruthy(item)) {
                    found = true;
                    break;
                }
            }
            var condition = context.transformItem(by, item, i++);
            if (adapter.isTruthy(condition)) {
                found = true;
                break;
            }
        }
        return found ? i - 1 : -1;
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionFindIndex.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import { isTruthy } from "../JsonHelpers";

class TransformerFunctionFindIndex extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        by: { type: ArgType.Transformer, position: 0, defaultIsNull: true },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const streamer = await context.getJsonElementStreamer(null);
    if (streamer == null) return null;

    const hasBy = context.has("by");
    const by = await context.getJsonElement("by", false);
    let index = 0;
    return await streamer.stream().indexOfFirst(async item => {
      if (!hasBy) {
        return isTruthy(item);
      }
      const condition = await context.transformItem(by, item, index++);
      return isTruthy(condition);
    });
  }
}

export default TransformerFunctionFindIndex;
